### PR TITLE
Share code amongst "Chrome-like" browsers

### PR DESCRIPTION
**Pull request by saghul**

### CODE REVIEW

**Comment by saghul** [comment]
Jenkins please test this please.

File: modules/RTC/RTCUtils.js [hunk 1]
```diff
@@ -814,10 +814,7 @@ class RTCUtils extends Listenable {
 
             this.getStreamID = ({ id }) => id;
             this.getTrackID = ({ id }) => id;
-        } else if (browser.isChrome() // this is chrome < 61
-                || browser.isOpera()
-                || browser.isNWJS()
-                || browser.isElectron()
+        } else if (browser.isChromeLike() // this is chrome < 61
```
Comment (bbaldino) [inline modules/RTC/RTCUtils.js]: is this comment still accurate?

File: modules/browser/BrowserCapabilities.js [hunk 2]
```diff
@@ -20,6 +20,7 @@ export default class BrowserCapabilities extends BrowserDetection {
         super();
         logger.info(
             `This appears to be ${this.getName()}, ver: ${this.getVersion()}`);
+        this.usesNewGumFlow() && logger.debug('Using the new GUM flow');
```
Comment (bbaldino) [inline modules/browser/BrowserCapabilities.js]: seems like an odd place to log this, do we not have a log for this when we go down the new gum path?  if not i'd rather have it there than inside this util class

File: modules/browser/BrowserCapabilities.js [hunk 3]
```diff
@@ -47,6 +48,17 @@ export default class BrowserCapabilities extends BrowserDetection {
         return !this.isEdge() && !this.isFirefox();
     }
 
+    /**
+     * Checks if the current browser is "Chrome like", that is, it's either
+     * Chrome / Chromium or uses it as its engine.
+     */
+    isChromeLike() {
```
Comment (bbaldino) [inline modules/browser/BrowserCapabilities.js]: we have both this and https://github.com/jitsi/js-utils/pull/6 ?

File: modules/browser/BrowserCapabilities.js [hunk 4]
```diff
@@ -241,11 +241,32 @@ export default class BrowserCapabilities extends BrowserDetection {
      * @returns {boolean}
      */
     usesNewGumFlow() {
-        return (this.isChrome()
-                && !this.isVersionLessThan('61'))
-            || this.isFirefox()
-            || this.isSafariWithWebrtc();
+        const REQUIRED_VERSION = 61;
```
Comment (bbaldino) [inline modules/browser/BrowserCapabilities.js]: `REQUIRED_CHROME_VERSION`?

File: modules/browser/BrowserCapabilities.js [hunk 5]
```diff
@@ -241,11 +241,32 @@ export default class BrowserCapabilities extends BrowserDetection {
...
+            // eslint-disable-next-line no-undef
+            const version = Number.parseInt(process.versions.chromium, 10);
+
+            return version >= REQUIRED_VERSION;
+        }
+
+        if (this.isChromeLike()) {
+            const ua = navigator.userAgent;
+
+            if (ua.match(/Chrome/)) {
```
Comment (bbaldino) [inline modules/browser/BrowserCapabilities.js]: this also seems like overalp with https://github.com/jitsi/js-utils/pull/6 ?

File: modules/RTC/RTCUtils.js (hunk 1 above)
Comment (saghul) [reply modules/RTC/RTCUtils.js]: Yes: we use that code for anything which uses the Chrome engine and its version is < 61.

File: modules/browser/BrowserCapabilities.js (hunk 2 above)
Comment (saghul) [reply modules/browser/BrowserCapabilities.js]: Here is where we log the browser name, so I thought it was an appropriate place. I'm ok with dropping it altogether though, it was helpful when debugging.

Comment (bbaldino) [reply modules/browser/BrowserCapabilities.js]: ok, my vote would be to log it in the flow that decides whether or not to use the new gum flow, but won't hold up the PR on the change.  just seems a bit weird to call out this one behavior of the browser capabilities in a log message here when there are lots of different flows we make based on the version.

File: modules/browser/BrowserCapabilities.js (hunk 3 above)
Comment (bbaldino) [reply modules/browser/BrowserCapabilities.js]: I guess what I don't understand is if we already detect chrome-like things as chrome, that means this check is effectively the same as:

```
 isChromeLike() {
    return this.isChrome();
}
```
no?  electron/js/opera will already be detected as chrome-like and therefore chrome, no? so what is this adding?  or maybe i'm missing something.

File: modules/browser/BrowserCapabilities.js (hunk 5 above)
Comment (bbaldino) [reply modules/browser/BrowserCapabilities.js]: so we detect chrome-like things as chrome, but don't parse the chrome version but instead the actual product (electron, etc.) version?  this seems like it would get confusing/misleading, as we represent things as chrome but parse a version that isn't actually the chrome version?

Comment (saghul) [reply modules/browser/BrowserCapabilities.js]: No, that's not what happens. Sort of. When `isChrome` returns true, that browser is detected as Chrome, so the version is the Chrome version. This is the case for Chrome, Chromum, Brave and after my js-utils patch, Vivaldi.

Then we have Electron, NWJS and Opera for which `isChrome` returns false, and their version is the product version. These are "Chrome like" and we want to extract the Chrome version by hand. I'll add a comment which hopefully clarifies this.

Thanks for being so thorough!

**Comment by saghul** [comment]
@bbaldino Updated, PTAL. I clarified this with comments as much as I could. Hopefully it's clearer now!

File: modules/browser/BrowserCapabilities.js [hunk 6]
```diff
@@ -241,11 +248,40 @@ export default class BrowserCapabilities extends BrowserDetection {
...
+
+        if (this.isChrome()) {
+            return !this.isVersionLessThan(REQUIRED_CHROME_VERSION);
+        }
+
+        if (this.isFirefox() || this.isSafariWithWebrtc()) {
+            return true;
+        }
+
+        if (this.isNWJS()) {
```
Comment (bbaldino) [inline modules/browser/BrowserCapabilities.js]: why is this needed as a separate case from isChromeLike?  Is it because NWJS represents the chrome version differently than others?  If so, could that be handled somewhere else so it remains more uniform with the rest of the isChromeLike tests?

Comment (bbaldino) [reply modules/browser/BrowserCapabilities.js]: if it is just the version thing, i wonder if it might be clearer to still have this as a case within the `if (this.isChromeLike()) {` block below?

**Comment by saghul** [comment]
>  and i wonder if something like `isChromiumBased` would be clearer?

I think you're right. Thanks for the suggestion! Will fix.

File: modules/browser/BrowserCapabilities.js (hunk 6 above)
Comment (saghul) [reply modules/browser/BrowserCapabilities.js]: >  Is it because NWJS represents the chrome version differently than others?

Yep, it's not in `navigator.userAgent` but in `process`.

Comment (saghul) [reply modules/browser/BrowserCapabilities.js]: I don't think the extra indentation level adds much IMHO. That's why I left it separate. I'll give it a shot when I rename the function and see how it feels like.

**Comment by saghul** [comment]
@bbaldino Updated, PTAL!

**Comment by saghul** [comment]
Jenkins please test this please.

**Comment by saghul** [comment]
Note: merged with a failing iOS build. (0 tests failed).
